crates/polars-ops/src/series/ops/cum_agg.rs

```diff
@@ -216,15 +216,7 @@ pub fn cum_max(s: &Series, reverse: bool) -> PolarsResult<Series> {
 }
 
 pub fn cum_count(s: &Series, reverse: bool) -> PolarsResult<Series> {
-    if reverse {
-        let ca: NoNull<UInt32Chunked> = (0u32..s.len() as u32).rev().collect();
-        let mut ca = ca.into_inner();
-        ca.rename(s.name());
-        Ok(ca.into_series())
-    } else {
-        let ca: NoNull<UInt32Chunked> = (0u32..s.len() as u32).collect();
-        let mut ca = ca.into_inner();
-        ca.rename(s.name());
-        Ok(ca.into_series())
-    }
+    let s = s.is_not_null().cast(&IDX_DTYPE).unwrap();
+    let ca = s.idx().unwrap();
+    Ok(cum_sum_numeric(ca, reverse).into_series())
```

Comment:
This can be more efficient because we know there are no nulls in the chunked array - so we can write a dedicated cum_sum for non-null booleans here generating IdxSize values. Or maybe there's a better way even?

Reply:
I was looking at this early, and I was blocked by my lack of understanding of the traits and such involved, but I was thinking something along these lines, not sure if it would be any better (it follows the same logic as the other functions in `cum_agg.rs`):

```rust
fn det_cum_count<T>(state: &mut u32, v: Option<T>) -> Option<Option<u32>>
where
    T: Copy + PartialOrd + AddAssign + Add<Output = T>,
{
    match v {
        Some(v) => {
            Some(*state + 1)
        },
        None => Some(*state),
    }
}

pub fn cum_count(s: &Series, reverse: bool) -> PolarsResult<Series> {
    
    let init = 0u32;
    // <---not sure how to do this generically, I couldn't get `into_iter` working
    let ca: &ChunkedArray<T> = s.somehow_get_ca_that_can_iter();
    let out = match reverse {
        false => ca.into_iter().scan(init, det_cum_count).collect_trusted(),
        true => ca.into_iter().rev().scan(init, det_cum_count).collect_reversed(),
    };
    Ok(out.with_name(ca.name())).into_series()
}
```

Reply:
Hey @stinodego, here's a working (tested) `cum_count` for the basic dtypes that is probably pretty fast (might be faster? no clue). No worries if it's not as good, I'm mainly just practicing since my rust is pretty basic. It iterates through the underlying ChunkedArray and builds the cum count vector directly, incrementing the count by 1 when the value isn't null. As I said, my rust is still pretty basic so this could probably be condensed a bit with macros and other improvements.

```python
pub fn cum_count(s: &Series, reverse: bool) -> PolarsResult<Series> {
    if s.null_count() == 0 {
        // fast path: no nulls
        return Ok(Series::new(s.name(), 1u32..(s.len() + 1) as u32));
    } else if s.null_count() == s.len() || s.dtype() == &DataType::Null {
        // fast path: all nulls
        return Ok(Series::from_vec(s.name(), vec![0u32; s.len()]));
    }
    let mut count = 0u32;
    let mut out = Vec::<u32>::with_capacity(s.len());
    macro_rules! accum_vec {
        ($ca:expr) => {
            if reverse {
                for v in $ca.into_iter().rev() {
                    // only incr by 1 if not null
                    if let Some(_) = v {
                        count += 1;
                    }
                    out.push(count);
                }
            } else {
                for v in $ca.into_iter() {
                    if let Some(_) = v {
                        count += 1;
                    }
                    out.push(count);
                }
            }
        };
    }
    match s.dtype() {
        DataType::Boolean => accum_vec!(s.bool()?),
        DataType::UInt8 => accum_vec!(s.u8()?),
        DataType::UInt16 => accum_vec!(s.u16()?),
        DataType::UInt32 => accum_vec!(s.u32()?),
        DataType::UInt64 => accum_vec!(s.u64()?),
        DataType::Int8 => accum_vec!(s.i8()?),
        DataType::Int16 => accum_vec!(s.i16()?),
        DataType::Int32 => accum_vec!(s.i32()?),
        DataType::Int64 => accum_vec!(s.i64()?),
        DataType::Float32 => accum_vec!(s.f32()?),
        DataType::Float64 => accum_vec!(s.f64()?),
        DataType::String => accum_vec!(s.str()?),
        _ => polars_bail!(ComputeError: "Invalid dtype"),
    }

    if reverse {
        // is there from_vec reversed?
        Ok(Series::from_vec(s.name(), out.reverse()))
    } else {
        Ok(Series::from_vec(s.name(), out))
    }
}
```

Reply:
The fast paths are a good idea. But I don't want to make this generic for all dtypes - I don't think it's worth the compilation time.

Reply:
@stinodego is there a way to grab a series' null bitmap without caring what the dtype is? If so the `cum_count` function could use that and have a single implementation for all dtypes.